### ISSUE TITLE
When operating on stan::math::var, does the gradient information propagate? 

### ISSUE THREAD

**Issue by charlesm93** [stan-dev/stan#1685]
I am writing a library of built-in functions in C++, which I would like to call from a .stan file. 
These functions can take random variables, defined in the parameters blocks, as arguments. This means the output of the function will be random variables. I understand that both the arguments and the output should be of type stan::math::var. 

I would like to understand how the gradient information (the adjoint) contained in the arguments will propagate to the final output. 

Taking a simple example, if I have a function f(x) = 2x, do I need to compute the adjoint of f with respect to the distribution I am sampling from? Experimenting a little with stan::math::var I have found that for the adjoint to be computed, I need to apply the grad() function to the distribution, which must be explicitly expressed in terms of x. On the other hand, if I run an operation on x (say 2*), I found x.adj() goes to 0. 

How will the adjoint be computed for f(x)? Does Stan take care of it when it calculates the likelihood or must the adjoint be part of the output of my function?

**Comment by syclik** [comment]
@charlesm93:

1. can you post this on stan-users? https://groups.google.com/d/forum/stan-user
    This isn't the right place to have this conversation and I'd prefer to close the issue.
2. have you looked at the auto-diff paper? http://arxiv.org/abs/1509.07164
3. I don't follow what you're saying, so I'll need help understanding what you mean.

**Comment by charlesm93** [comment]
@syclik 
1. I applied to join the google group and will create a new topic when granted membership.
2. I went through the paper to the best of my ability.
3. Is there a specific part you would like me to clarify? Maybe another way of to phrase my question could be: I created a variable which depends on a random variable. The relationship between the two variables is expressed by a function I wrote in C++ and added to Stan's library. In the Stan file, my expression is variable2 = f(variable1). But when returned, I expect that variable2.adj() = 0. Should this be a concern, given the calculation of the likelihood should also depend on variable2?
